source/server/admin/prometheus_stats_request.cc

```diff
@@ -0,0 +1,160 @@
+#include "source/server/admin/prometheus_stats_request.h"
+
+#include <string>
+#include <vector>
+
+#include "source/server/admin/prometheus_stats.h"
+
+#include "stats_params.h"
+#include "stats_render.h"
+
+namespace Envoy {
+namespace Server {
+
+namespace {
+bool metricLessThan(const Stats::RefcountPtr<Stats::Metric>& stat1,
+                    const Stats::RefcountPtr<Stats::Metric>& stat2) {
+  ASSERT(&stat1->constSymbolTable() == &stat2->constSymbolTable());
+  return stat1->constSymbolTable().lessThan(stat1->statName(), stat2->statName());
+}
+} // namespace
+
+PrometheusStatsRequest::PrometheusStatsRequest(Stats::Store& stats, const StatsParams& params,
+                                               Stats::CustomStatNamespaces& custom_namespaces,
+                                               UrlHandlerFn url_handler_fn)
+    : StatsRequestBase(stats, params, url_handler_fn), custom_namespaces_(custom_namespaces) {
+
+  // the "type" query param is ignored for prometheus stats, so always start from
+  // counters; also, skip the TextReadouts phase unless that stat type is explicitly
+  // requested via query param
+  if (params_.prometheus_text_readouts_) {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::TextReadouts, "Text Readouts"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  } else {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  }
+  phase_ = 0;
+}
+
+template <class StatType> Stats::IterateFn<StatType> PrometheusStatsRequest::checkStat() {
+  return [this](const Stats::RefcountPtr<StatType>& stat) -> bool {
+    // check if unused
+    if (params_.used_only_ && !stat->used()) {
+      return true;
+    }
+
+    // check if filtered
+    if (params_.filter_ != nullptr) {
+      if (!std::regex_search(stat->name(), *params_.filter_)) {
```

Comment:
I'm wondering if filters are ever used with Prom. It seems weird now that I think of it to filter on the fully elaborated name, but I guess that's easier from a coding and usage perspective than filtering on the tag-extracted name and then having a separate mechanism for filtering on the extracted tags.